Keeps preview across different tabs

WiNloSt opened the pull request:
Part of a milestone 6 of
Epic: https://github.com/metabase/metabase/issues/35961

### Description

Make sure embed preview doesn't reload when changing between tabs/preview toggle/appearance settings

Please note that when going to the overview tab, the preview will be unmounted, so going to the preview again after visiting the overview tab will cause the iframe to reload. This is a compromise for the current codebase. I need to put a lot more work on the overview tab to make that happen, so it didn't happen.

### How to verify
Tests both dashboards and questions the same way.
1. Open static embed modal for a dashboard/question
2. Go to Parameters tab.
3. Click preview
4. Go to Appearance tab. The preview shouldn't rerender.
5. Modify settings in the Appearance tab, the preview shouldn't rerender.

### Demo

https://www.loom.com/share/f700c55d53634f5b974bba30f4f55867

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

WiNloSt committed:
Keep preview when toggling between code/preview in each tab

WiNloSt committed:
Persist preview across tabs

WiNloSt committed:
[WIP] don't reload embed preview when changing appearance

TODO:
- Fix font loading

WiNloSt committed:
Merge branch 'master' into 35961-keep-preview-across-different-tabs

WiNloSt committed:
Allow changing Static/public embedding font on the fly

WiNloSt committed:
Merge branch 'master' into 35961-keep-preview-across-different-tabs

WiNloSt committed:
Use 2-way dashboard state/URL hash syncing to avoid failing tests

WiNloSt committed:
Fix E2E tests after dashboard preview doesn't reload on every change

WiNloSt committed:
Use an existing util

WiNloSt committed:
Fix test name

WiNloSt committed:
Add tests to ensure preview doesn't rerender unnecessarily

WiNloSt commented on frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/PreviewPane.styled.tsx:
```diff
@@ -1,9 +1,20 @@
+  ${({ hidden }) =>
+    hidden &&
+    css`
+      visibility: hidden;
```
Can't use `display: none` because it will cause the iframe to be reloaded again. So I need the iframe to stay in the dom. I set `position: absolute` so it doesn't take up space inside the modal when switching to other tabs.

WiNloSt commented on frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/StaticEmbedSetupPane.tsx:
```diff
@@ -107,19 +113,17 @@ export const StaticEmbedSetupPane = ({
   });
 
-  const iframeUrl = useMemo(
+  const iframeUrlWithoutHash = useMemo(
```
I need the URL to stay consistent when only the appearance option is changed. Otherwise, it will cause the iframe to rerender.

npretto replied:
Isn't this only used in `iframeUrl` which also uses the hash params?

WiNloSt replied:
yeah, this variable is used down below `const iframeUrl = iframeUrlWithoutHash + optionsToHashParams(displayOptions);`

Basically, I need to base URL to stay consistent, so I separate the calculation of base URL and hash. So if only the hash changes, the base URL would stay the same.

WiNloSt commented on frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/StaticEmbedSetupPane.tsx:
```diff
@@ -174,48 +183,90 @@ export const StaticEmbedSetupPane = ({
-            onChangePane={setActivePane}
-            onChangeDisplayOptions={setDisplayOptions}
-          />
+        <Tabs.Panel value={activeTab}>
```
This is a bit hacky, but essentially, I need the preview to stay in the same component, otherwise, it will unmount and remount when changing between tabs because we have 3 parents (`Tabs.Panel`) previously. Now it's reduced to just 1.

npretto replied:
It took a bit to understand why this is working lol, even with the context of the pr and your comments, should we add a comment? Otherwise it could be very easy to think "why are they doing this??" when reading the code :D

WiNloSt replied:
Yeah, good point, the comment should go into the code.

WiNloSt replied:
Added a comment in https://github.com/metabase/metabase/pull/38271/commits/2b8172c10d0c10cf15f6aa25665b869f508ac84d

WiNloSt commented on frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/StaticEmbedSetupPane.tsx:
```diff
@@ -174,48 +183,90 @@ export const StaticEmbedSetupPane = ({
-            onChangeDisplayOptions={setDisplayOptions}
-          />
+        <Tabs.Panel value={activeTab}>
+          {activeTab === TABS.Overview ? (
```
Ternary isn't the best looking thing here, but I need every element to be in the same bracket, otherwise the iframe preview will rerender.

WiNloSt commented on frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/StaticEmbedSetupPane.tsx:
```diff
@@ -174,48 +183,90 @@ export const StaticEmbedSetupPane = ({
-          />
+        <Tabs.Panel value={activeTab}>
+          {activeTab === TABS.Overview ? (
+            <OverviewSettings
```
I wish I could keep the preview state even when navigating to the Overview tab, but it's a bit hard to extract the content inside it here, so we only keep the embedding state when switching between Parameters and Appearance tabs.

WiNloSt commented on frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/StaticEmbedSetupPane.tsx:
```diff
@@ -174,48 +183,90 @@ export const StaticEmbedSetupPane = ({
+              }
+              previewSlot={
+                <>
+                  <PreviewModeSelector
```
This is moved from both `ParametersSettings` and `AppearanceSettings`, the structures are similar that's why React doesn't unmount the preview when changing between these 2 tabs.

WiNloSt commented on frontend/src/metabase/public/lib/embed.ts:
```diff
@@ -29,12 +28,11 @@ function getSignedToken(
 }
 
-export function getSignedPreviewUrl(
+export function getSignedPreviewUrlWithoutHash(
```
This is only used in `frontend/src/metabase/public/components/EmbedModal/StaticEmbedSetupPane/StaticEmbedSetupPane.tsx` so it's safe to modify the return value shape. I removed the hash part because I want the URL part before the hash to stay the same when only the hash is changed.

With the previous implementation, even when the hash is changed, the whole URL will be changed too.

npretto commented on e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js:
```diff
@@ -423,6 +426,94 @@ describe("scenarios > embedding > dashboard parameters with defaults", () => {
+      cy.log("Assert dashboard title");
+      getIframeBody().findByText(dashboardDetails.name).should("exist");
+      // We're getting an input element which is 0x0 in size
+      cy.findByLabelText("Dashboard title").click({ force: true });
```
Instead of finding the element with that label, can we click the label?

WiNloSt replied:
That's a good point. Do you have an example where we click the labels? I looked up testing-library, and they could only get the input using various matchers e.g. `byPlaceholderText`, `byLabelText`, but to find labels I could use `findByText` which is less specific, or `findByRole` with a lot of hoops. I'd rather use `findByLabelText` with `force: true` because I think it's still more readable.

What do you think?

npretto reviewed (approved):
Added some comments but looks good to me

WiNloSt committed:
Merge branch 'master' into 35961-keep-preview-across-different-tabs

WiNloSt committed:
Fix dashboard embed preview reload when it shows parameters

WiNloSt committed:
Review: Comment code for clarity

WiNloSt committed:
Fix tests after fixing the preview In Cypress

WiNloSt commented on e2e/test/scenarios/embedding/embedding-dashboard.cy.spec.js:
```diff
@@ -423,6 +426,112 @@ describe("scenarios > embedding > dashboard parameters with defaults", () => {
+      type: "string/contains",
+    };
+
+    const dashboardDetails = {
```
@npretto I added this filter, since I found out today that having a parameter will cause the appearance page preview to rerender every time we change any settings

WiNloSt commented on e2e/test/scenarios/embedding/reproductions/37914-embed-preview-parameter-values-not-working.cy.spec.js:
```diff
@@ -52,7 +53,9 @@ const dashboardDetails = {
 describe("issue 37914", () => {
   beforeEach(() => {
-    cy.intercept("GET", "/api/embed/dashboard/**").as("getEmbed");
+    cy.intercept("GET", "/api/preview_embed/dashboard/**").as(
```
The preview didn't seem to work in Cypress before. I've fixed that in `frontend/src/metabase/lib/dom.js`

WiNloSt commented on e2e/test/scenarios/sharing/reproductions/26988-embedding-dynamic-settings.cy.spec.js:
```diff
@@ -40,7 +42,7 @@ describeEE("issue 26988", () => {
     });
 
-    cy.wait("@dashboard");
+    cy.wait("@previewDashboard");
```
The preview didn't seem to work in Cypress before. I've fixed that in `frontend/src/metabase/lib/dom.js`

WiNloSt commented on frontend/src/metabase/lib/dom.js:
```diff
@@ -26,7 +26,7 @@ window.METABASE = true;
 export const IFRAMED_IN_SELF = (function () {
   try {
-    return window.self !== window.top && window.top.METABASE;
+    return window.self !== window.parent && window.parent.METABASE;
```
`IFRAMED_IN_SELF` was only used in a few places. And all of them relate to how we're rendering the embed preview.

This was failing in Cypress because there are 3 frames in Cypress
- Cypress
    - Metabase Iframe
        - Embed preview Iframe

But in our application, there are only 2 frames
- Metabase
    - Embed preview Iframe
 
For this particular code, when running in Cypress `window.parent` will get Cypress frame, so the function will return `false`, which was wrong.

Changing it to `window.parent`, makes sure it returns the immediate parent frame which will be Metabase frame in both cases.

WiNloSt commented:
@npretto Can you help review the last 3 commits https://github.com/metabase/metabase/pull/38271/files/fe2a43912ed59d37bfbe77ecad5d9433042dfb1b..88ce7d07921a3c1c4bab7be362aac29bdd62570d?

I've fixed one more scenario and modified the test to capture that.

albertoperdomo reviewed (approved):
LGTM

WiNloSt committed:
Merge branch 'master' into 35961-keep-preview-across-different-tabs
